packages/swapper/src/swappers/cow/utils/helpers/helpers.ts

```diff
@@ -0,0 +1,61 @@
+import { fromAssetId } from '@shapeshiftoss/caip'
+import { Asset, ChainTypes, NetworkTypes } from '@shapeshiftoss/types'
+import { AxiosResponse } from 'axios'
+
+import { SwapError, SwapErrorTypes } from '../../../../api'
+import { bn, bnOrZero } from '../../../utils/bignumber'
+import { CowSwapperDeps } from '../../CowSwapper'
+import { CowSwapPriceResponse } from '../../types'
+import { cowService } from '../cowService'
+
+const USDC_CONTRACT_ADDRESS = '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48'
+
+export const getUsdRate = async (
+  { assetService, apiUrl }: CowSwapperDeps,
+  input: Pick<Asset, 'assetId'>
+): Promise<string> => {
+  const { assetId } = input
+
+  const { assetReference: erc20Address, assetNamespace } = fromAssetId(assetId)
+
+  if (assetNamespace !== 'erc20') {
+    throw new SwapError('[getUsdRate] - unsupported asset namespace', {
+      code: SwapErrorTypes.USD_RATE_FAILED,
+      details: { assetNamespace }
+    })
+  }
+
+  const asset = assetService.byTokenId({
+    chain: ChainTypes.Ethereum,
+    network: NetworkTypes.MAINNET,
+    tokenId: erc20Address
+  })
+
+  try {
+    // rate is imprecise for low $ values, hence asking for $1000
```

Comment:
maybe best to define 1000 in a named const for clarity and derive all values from that (its not clear what 1000000000 represents for example)

Reply:
> How is it imprecise for low values? That might be ok? $1000 feels like an arbitrary value

I chose $1000 through trial and testing, I corrected similar bug couple months ago for Zrx where getUsdRate() was returning very imprecise values.
Extracted 1000 to a const for more clarity !

Reply:
Can you give me an example of the price difference? im curious now

I remember seeing that code in zrx swapper when we were cleaning it up and I didnt feel like it was necessary for zrx because its an AMM and large amounts shouldnt make much different. 

Buit maybe its more important for cowswap because it uses an order book and small prices can be very far off? I just want to make sure I fully understand the implications of changing this amount

Reply:
@elmutt 
One example with cowswap : 

WBTC value currently : **$29522**

** usdRate fetching $1

Api call : https://api.cow.fi/mainnet/api//v1/markets/0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48-0x2260fac5e5542a773aa44fbcfedf7c193bc2c599/buy/1000000

Api response : {"amount":"3277","token":"0x2260fac5e5542a773aa44fbcfedf7c193bc2c599"}
usdRate calculation leads to **WBTC 30515.71**

** usdRate fetching $1000
Api call : https://api.cow.fi/mainnet/api//v1/markets/0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48-0x2260fac5e5542a773aa44fbcfedf7c193bc2c599/buy/1000000000

Api response : {"amount":"3389831","token":"0x2260fac5e5542a773aa44fbcfedf7c193bc2c599"}
usdRate calculation leads to **WBTC 29499.99**

I witnessed same behaviour with multiple tokens (FOX etc).

CowSwap ultimately will look into 0x if needed so there is a good chance the same behaviour on 0x is simply present on CowSwap

Reply:
I dont think its a "bug" for 0x to use an amount of $1 because the price is always going to be different depending on the amount. Im think for 0x using $1000 is just as arbitrary as $1 -- It will ee wrong if they do am amount higher or lower than $1000. The price would be correct for a trade amount of X but its likely the user does a different amount than we are using to estimate the price.

Ultimately usdRate is only used as an estimate and its impossible for to be perfect without specifying an amount. Using $1000 here is fine if you feel it gets us the closest estimate


Its strange to me that the cowswap rate gets better as amount increases, I would expect it to be the opposite because of slippage. Do you know what that is?

Reply:
I agree with you that $1 is not any more "right" than $1000, but ultimately some api calls will be more relevant regarding the usdRate estimation we want, and that's what I looked for ...

For the rate getting better as amount increase, it's really an effect I noted going from $1 to $1000 (for both 0x and cowswap). I would say slippage is minimal for those "low" values of liquid coins like WBTC (also it seems slippage is not taken into account in this api from their doc : https://docs.cow.fi/off-chain-services/api/price-estimation)

I am not sure but I think it's more a fees problem.  Since fees are taken into account with the amount of token we can expect from a buy, it may make sense that fee represent a very large percentage of the transaction value for tiny values on ETH network, hence this bad estimation we are seeing with low value orders. 

I will certainly get more information as I dig more into the cowswap api ...